### PR TITLE
Fix Wireshark dissector and add snapshot tests

### LINKED ISSUE #726: Wireshark dissect bit fields fails
Most notably this affects HEARTBEAT dissection.

looks like this change https://github.com/ArduPilot/pymavlink/commit/4a8b143c071ee2e937bba36411ba1f8fb9c1a7ab
broke the bit field dissection in wireshark.
The line 
   _subtree, value = tree:add_le(f.${fvar}, padded(field_offset, ${fbytes}))_
in generated lua does not actually return the value, it returns nil instead.
here is a generated example
    _subtree, value = tree:add_le(f.HEARTBEAT_base_mode, padded(field_offset, 1))_

As a work around I explicitly extracted the value to pass to the generated dissect_flags function which resolved the issue.
in mavgen_wlua.py, function generate_field_dissector. line 319

```
        if flag_enum is not None:
            t.write(outf,
"""
    value = padded(field_offset, ${fbytes}):uint()
    dissect_flags_${enumname}(subtree, "${fvar}", value)
""", {'enumname': flag_enum.name, 'fvar': field_var, 'fbytes': size})

```

Of course this is assuming all bit fields are an uint

### LINKED ISSUE #815: Wireshark Wlua does not dissect flags
An error is shown if you try to inspect a message, such as HEARTBEAT, that has flags, and you can't inspect any fields after that flag. For example in HEARTBEAT this is the mode flag:

There are two issues. Firstly, the `value` returned by `tree:add_le` is `nil`, so what gets passed into `dissect_flags_MAV_MODE_FLAG` cannot be dissected. The docs say `tree:add_le` should return a value. My assumption is that this is failing for the reason given after the code. Note that this is the only place the value is actually used, which is why we don't see the error elsewhere.

```
    field_offset = offset + 6
    subtree, value = tree:add_le(f.HEARTBEAT_base_mode, padded(field_offset, 1))
    dissect_flags_MAV_MODE_FLAG(subtree, "HEARTBEAT_base_mode", value)
```

I "fixed" this error by passing in `padded(field_offset, 1)` to `dissect_flags_MAV_MODE_FLAG` instead of `value`. Wireshark now fails with the error that the data passed in is "userData" - i.e. an arbitrary C data that it can't interpret. That's correct, and probably why `nil` got returned from the `tree:add_le`.

The solution would be to cast the data to a number before passing it to `tree:add_le`. HOwever I don't know how to do this elegantly.

This is cross posted (and summarised) from https://github.com/mavlink/mavlink/issues/1978

### PR DESCRIPTION
**Pull request by rotu**
Handle bit flags in Wireshark dissector (fix #815, fix #726)
Add tests for Wireshark (via `tshark` binary and `pytest` via the `syrupy` snapshot library)
Remove defunct mav2pcap.py script

### CODE REVIEW

**Comment by hamishwillee** [comment]
@rotu Tried this out. Certainly fixes the error we were seeing. I particularly like how clear the bitflags are rendered:

image

I also looked at many other message, including commands, and nothing seems broken:

image

Excellent to see test code of this. Would be good to expand it in future as a validation test. Out of scope for this probably.

The generator code changes make sense to me, and I think this also makes the generated code much more clear. 

However I am not expert. Once you have this to a state you are happy with, we'll need to  work with @peterbarker to get this in.

**Comment by rotu** [comment]
I'm not an expert either. Thanks for the screenshots! I think they highlight some other bugs!

The second message you show is a [`COMMAND_INT`](https://mavlink.io/en/messages/common.html#COMMAND_INT), where param2 is a float.

The value shown is `1065353216` which I think is supposed to be the floating point value `1.0f`, corresponding to `1`; i.e. `MAV_DO_REPOSITION_FLAGS_CHANGE_MODE: True`.

Also the longitude/latitude values are payload parameters 5/6 so I think they should be `int32_t`, not `float`, hence the wonky-looking values.

Could you please attach some pcap data I can test against?

**Comment by hamishwillee** [comment]
Looks like you're right. I ran this for a few seconds and then did a save-as. Is this what you needed?

[pcapdata.zip](https://github.com/ArduPilot/pymavlink/files/11548904/pcapdata.zip)

**Comment by hamishwillee** [comment]
PS Also a CI error

**Comment by rotu** [comment]
I see the CI errors and but the alternative is to use the [obsolete `pkg_resources`](https://setuptools.pypa.io/en/latest/pkg_resources.html) and I hate writing new legacy code. When are we dropping python 2.7 and 3.6 support per #556?

**Comment by hamishwillee** [comment]
@rotu Could not agree more. 
PX4 have dropped support for Python 2, and I understand that ArduPilot either have done so, or will be doing so very shortly. If those dependencies are gone, we should be able to switch over "very soon". I'll push for that, so ignore the error for now. 
Wr.t. dropping Python 3.6 that is harder - we are in discussion.

As I understand it you're looking at the other issues with the generated output. Since I'm very happy for you to do so I'm going to just keep quiet now :-)

**Comment by hamishwillee** [comment]
@rotu This isn't waiting on me for anything is it?

**Comment by hamishwillee** [comment]
Hi @rotu 
Looks like you are still progressing. Let me know when you need more review.

**Comment by rotu** [comment]
@hamishwillee Yep. I'm chipping away slowly. I'm a little hung up on getting the output to match when there are long lines (not sure how to get the output of tshark to wrap to a consistent line length)

**Comment by rotu** [comment]
Alright I'm sufficiently sick of this PR, so it's time to get it out there!

**Comment by hamishwillee** [comment]
@rotu I can imagine. Time for me to get a technical review on this?

**Comment by rotu** [comment]
> @rotu I can imagine. Time for me to get a technical review on this?

I don't understand what you mean. Do whatever you need, loop in whoever you like, and ask whatever questions to satisfy your curiosity.

**Comment by hamishwillee** [comment]
@peterbarker Can you review this please. I think it is important to get in.

I have done basic tests on it by running the instructions in https://mavlink.io/en/guide/wireshark.html and also specifically looking at the HEARTBEAT message to check that the known issue was fixed. 
This definitely fixes the original problem in a logic way, and the resulting output is also good.

@rotu Has also added a heap of test code. I have not tested that, since I have no merge rights, and you will have to do this anyway.

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,7 @@
 lxml>=3.6.0
 future>=0.15.2
 wheel>=0.37.1
+
+# dev dependencies:
+pytest
```
Comment (peterbarker) [inline requirements.txt:6]: Should we add these dependencies for everyone on every install of pymavlink?

The vast majority of people won't want to run the test suite.

Perhaps we should add this as a requirements.txt in the `snapshottests` directory?

Comment (rotu) [reply requirements.txt:6]: > Should we add these dependencies for everyone on every install of pymavlink?

No, and this PR doesn't. When you `pip install` pymavlink, you *don't* get the stuff in requirements.txt. You get the (totally unversioned :shrug:) dependencies in `setup.py`.

Comment (hamishwillee) [reply requirements.txt:6]: @peterbarker Response please.

Comment (hamishwillee) [reply requirements.txt:6]: @peterbarker Ping.

Comment (peterbarker) [reply requirements.txt:6]: Cool, thanks.

**Review by peterbarker — CHANGES_REQUESTED** [review]
@rotu could you fix the commit list, please?  Should be a single commit

**Comment by rotu** [comment]
> ... should have said I'm happy enough to merge this with a good commit list.

@peterbarker
Feel free to squash merge. https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

**Comment by hamishwillee** [comment]
Hurray. Thanks!

**Pull request merged**